core/src/main/java/org/infinispan/remoting/transport/jgroups/JGroupsTransport.java

```diff
@@ -466,700 +626,845 @@ private void buildChannel() {
+      if (noRelay) {
+         message.setFlag(Message.Flag.NO_RELAY.value());
+      }
+      short flags = encodeDeliverMode(deliverOrder);
+      message.setFlag(flags);
+      // Only the commands in total order must be received by the originator.
+      if (deliverOrder != DeliverOrder.TOTAL) {
+         message.setTransientFlag(Message.TransientFlag.DONT_LOOPBACK.value());
+      }
+      if (rsvp) {
+         message.setFlag(Message.Flag.RSVP.value());
       }
-      return retval;
    }
 
-   @Override
-   public BackupResponse backupRemotely(Collection<XSiteBackup> backups, XSiteReplicateCommand rpcCommand) throws Exception {
-      if (trace) {
-         log.tracef("About to send to backups %s, command %s", backups, rpcCommand);
+   private void logCommand(ReplicableCommand command, long requestId, String targets) {
+      if (requestId != Request.NO_REQUEST_ID) {
+         log.tracef("Sending request %d@%s to %s: %s", requestId, address, targets, command);
+      } else {
+         log.tracef("Sending command from %s to %s: %s", address, targets, command);
       }
-      Buffer buf = dispatcher.marshallCall(rpcCommand);
-      Map<XSiteBackup, Future<Object>> syncBackupCalls = new HashMap<>(backups.size());
-      for (XSiteBackup xsb : backups) {
-         SiteMaster recipient = new SiteMaster(xsb.getSiteName());
-         if (xsb.isSync()) {
-            RequestOptions sync = constructRequestOptions(org.jgroups.blocks.ResponseMode.GET_ALL,
-                  false, DeliverOrder.NONE, xsb.getTimeout(), false);
-            syncBackupCalls.put(xsb, dispatcher.sendMessageWithFuture(recipient, buf.getBuf(), buf.getOffset(), buf.getLength(), sync));
+   }
+
+   private void send(Message message) {
+      try {
+         channel.send(message);
+      } catch (Exception e) {
+         if (channel.isConnected()) {
+            throw new CacheException(e);
          } else {
-            RequestOptions async = constructRequestOptions(org.jgroups.blocks.ResponseMode.GET_NONE,
-                  false, DeliverOrder.PER_SENDER, xsb.getTimeout(), false);
-            dispatcher.sendMessage(recipient, buf.getBuf(), buf.getOffset(), buf.getLength(), async);
+            throw log.cacheManagerIsStopping();
          }
       }
-      return new JGroupsBackupResponse(syncBackupCalls, timeService);
    }
 
-   private static org.jgroups.blocks.ResponseMode toJGroupsMode(ResponseMode mode) {
-      switch (mode) {
-         case ASYNCHRONOUS:
-            return org.jgroups.blocks.ResponseMode.GET_NONE;
-         case WAIT_FOR_VALID_RESPONSE:
-            return org.jgroups.blocks.ResponseMode.GET_FIRST;
-         case SYNCHRONOUS:
-         case SYNCHRONOUS_IGNORE_LEAVERS:
-            return org.jgroups.blocks.ResponseMode.GET_ALL;
+   private void addRequestHeader(Message message, long requestId) {
+      // TODO Remove the header and store the request id in the buffer
```

Comment:
Mostly to avoid allocating a header instance and reserving a header id.

But you're right, creating our own header is also an option, please comment on https://issues.jboss.org/browse/ISPN-7933 with your arguments for it.